src/jit/codegenxarch.cpp

```diff
@@ -176,23 +176,63 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
 
     // Make sure that EAX is reported as live GC-ref so that any GC that kicks in while
     // executing GS cookie check will not collect the object pointed to by EAX.
-    if (!pushReg && (compiler->info.compRetType == TYP_REF))
-        gcInfo.gcRegGCrefSetCur |= RBM_INTRET;    
+    if (!pushReg)
+    {
+
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+
+        if (varTypeIsStruct(compiler->info.compRetType))
+        {
+            if (compiler->compMethodReturnsMultiRegRetType())
+            {
+                // Get the return tye of the struct.
+                SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR retStructDesc;
+                compiler->eeGetSystemVAmd64PassStructInRegisterDescriptor(compiler->info.compMethodInfo->args.retTypeClass, &retStructDesc);
+
+                assert(retStructDesc.passedInRegisters);
+                // In case of a single register returned struct , the return type have been normalized.
+                // In such case the execution will not be here 
+                // (the if (varTypeIsStruct(compiler->info.compRetType)) will evaluate to false).
+                assert(retStructDesc.eightByteCount > 1 &&
+                    retStructDesc.eightByteCount <= CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_PASS_IN_REGISTERS);
+
+                unsigned __int8 offset0 = 0;
+                unsigned __int8 offset1 = 0;
+
+                var_types type0 = TYP_UNKNOWN;
+                var_types type1 = TYP_UNKNOWN;
+
+                // Set the GC-ness of the struct return registers.
+                getStructTypeOffset(retStructDesc, &type0, &type1, &offset0, &offset1);
+                gcInfo.gcMarkRegPtrVal(REG_INTRET, type0);
+                gcInfo.gcMarkRegPtrVal(REG_INTRET_1, type1);
+            }
+        }
+        else
+
+#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
+
+        {
+            gcInfo.gcMarkRegPtrVal(REG_INTRET, compiler->info.compRetType);
```

Comment:
Here we should be using compRetNativeType which is the normalized type in case of structs returned in a single register.